[RF-DOCS]Active Record Validations [ci skip] (#52951)
rails/rails: guides/source/active_record_validations.md

Active Record Validations
=========================

This guide teaches you how to validate Ruby objects before saving them to the
database using Active Record's validations feature.

After reading this guide, you will know:

* How to use the built-in Active Record validation helpers and options.

> * Checking the validity of objects.

Reviewer: ```suggestion
* How to check the validity of objects.
```

```ruby
validates_presence_of :name
```

Both notations perform the same function, but the newer form is recommended for
its readability and alignment with Rails' conventions.

Each validation accepts an arbitrary number of attribute names, allowing you to
apply the same type of validation to multiple attributes in a single line of
code.

Additionally, all validations accept the `:on` and `:message` options. The `:on`
option specifies when the validation should be triggered, with possible values
being `:create` or `:update`. The `:message` option allows you to define a
custom error message that will be added to the errors collection if the
validation fails. If you do not specify a message, Rails will use a default
error message for that validation.

INFO: To see a list of the available default helpers, take a look at
[`ActiveModel::Validations::HelperMethods`][]. This API section uses the older
notation as described above.

[`ActiveModel::Validations::HelperMethods`]:
    https://api.rubyonrails.org/classes/ActiveModel/Validations/HelperMethods.html

Below we outline the most commonly used validations.

### `absence`

> This helper validates that the specified attributes are absent. It uses the

Reviewer 2: I suppose that the following should be described first for newbies:

```diff
- This helper validates that the specified attributes are absent.
+ This helper validates that the specified attributes or an association are absent.
```

I suppose that the following (passing an association and attributes simultaneously) should also be clarified for newbies:

```ruby
class Supplier < ApplicationRecord
  has_one :account
  validates :account, :name, :login, :email, absence: true # possible? impossible? unrecommended?
end
```

> This helper validates that the specified attributes are absent. It uses the

Reviewer 3: ```suggestion
This validator validates that the specified attributes are absent. It uses the
```

[`Object#present?`][] method to check if the value is neither nil nor a blank
string - that is, a string that is either empty or consists of whitespace only.

```ruby
class Person < ApplicationRecord
  validates :name, :login, :email, absence: true
end

> ```

Reviewer 2: I'd be glad if you add some more working sample query methods below each of the sample models for `presence` and `absence`, to show how they work to newbies. 🙏

Reviewer 3: It's probably useful for conditional validations.
```ruby
class 
  validate :error_message, absence: true, if: :completed?
end
```

Author: > I'd be glad if you add some more working sample query methods (via IRB) as well, below each of the sample models for presence and absence, to show how they work to newbies. 🙏

This is a good idea, thanks! I've added some irb samples.

Reviewer 2: Thnx! 😂

If you want to be sure that an association is absent, you'll need to test
whether the associated object itself is absent, and not the foreign key used to
map the association. Rails will usually infer the inverse association automatically.

```ruby
class LineItem < ApplicationRecord

>   belongs_to :order

Reviewer 3: Or maybe replace the example with a `has_one` instead of the `belongs_to` as a `has_one` isn't required by default?

```ruby
class Supplier < ApplicationRecord
  has_one :account
  validates :account, presence: true
end

class Account < ApplicationRecord
  belongs_to :supplier
end
```

  validates :order, absence: true
end
```

> 

Reviewer 3: Might be good to mention the association presence is validated by default for `belongs_to`:
```suggestion

NOTE: For `belongs_to` the association presence is validated by default. If you don’t want to have association presence validated, use `optional: true`.

```

option specifies when the validation should be triggered, with possible values
being `:create` or `:update`. The `:message` option allows you to define a
custom error message that will be added to the errors collection if the
validation fails. If you do not specify a message, Rails will use a default
error message for that validation.

INFO: To see a list of the available default helpers, take a look at
[`ActiveModel::Validations::HelperMethods`][]. This API section uses the older
notation as described above.

[`ActiveModel::Validations::HelperMethods`]:
    https://api.rubyonrails.org/classes/ActiveModel/Validations/HelperMethods.html

Below we outline the most commonly used validations.

### `absence`

This validator validates that the specified attributes are absent. It uses the
[`Object#present?`][] method to check if the value is neither nil nor a blank
string - that is, a string that is either empty or consists of whitespace only.

`#absence` is commonly used for conditional validations. For example:

```ruby
class Person < ApplicationRecord
  validates :phone_number, :address, absence: true, if: :invited?
end
```

```irb

> person = Person.new(name: "Jane Doe", invitation_sent_at: Time.current)
> person.valid?

Reviewer 3: Without `irb>` the styling is a bit off:
<img width="875" alt="image" src="https://github.com/user-attachments/assets/a9daa322-24e1-4775-8f48-9846e37150e1">
```suggestion
irb> person = Person.new(name: "Jane Doe", invitation_sent_at: Time.current)
irb> person.valid?
```

=> true # absence validation passes
```

If you want to be sure that an association is absent, you'll need to test
whether the associated object itself is absent, and not the foreign key used to
map the association.

```ruby
class LineItem < ApplicationRecord
  belongs_to :order, optional: true
  validates :order, absence: true
end
```

```irb

> line_item = LineItem.new
> line_item.valid?

Reviewer 3: ```suggestion
irb> line_item = LineItem.new
irb> line_item.valid?
```

=> true # absence validation passes

order = Order.create

> line_item_with_order = LineItem.new(order: order)
> line_item_with_order.valid?

Reviewer 3: ```suggestion
irb> line_item_with_order = LineItem.new(order: order)
irb> line_item_with_order.valid?
```

```

```irb
irb> new_person = Person.new
irb> new_person.errors[:name]
=> [] # no errors since validations are not run until saved
irb> new_person.errors[:name].any?
=> false

irb> saved_person = Person.create
irb> saved_person.errors[:name]
=> ["can't be blank"] # validation error because `name` is required
irb> saved_person.errors[:name].any?
=> true
```

Additionally, you can use the
[`errors.add`](https://api.rubyonrails.org/classes/ActiveModel/Errors.html#method-i-add)
method to manually add error messages for specific attributes. This is
particularly useful when defining custom validation scenarios.

```ruby
class Person < ApplicationRecord
  validate do |person|
    errors.add :name, :too_short, message: "is not long enough"
  end
end
```

NOTE: To read about validation errors in greater depth refer to the [Working
with Validation Errors](#working-with-validation-errors) section.

[Errors#squarebrackets]:
    https://api.rubyonrails.org/classes/ActiveModel/Errors.html#method-i-5B-5D

Validation Helpers

> ------------------

Author: This is an excellent point. I did the following: 

1. I renamed the header to Validations 
2. I added this section to clarify the different notations:

> In modern Rails applications, the more concise validate syntax is commonly used, for example:
> 
> ```
> validate :name, presence: true
> ```
> 
> However, older versions of Rails used "helper" methods, such as:
> 
> ```
> validates_presence_of :name
> ```
> 
> Both notations perform the same function, but the newer form is recommended for its readability and alignment with Rails' conventions.

3. In the info section that links to the API Helpers

> INFO: To see a list of the available default helpers, take a look at
> [`ActiveModel::Validations::HelperMethods`][]. This API section uses the older notation as described above.

4. User "validations" instead of "validation helpers".

See https://github.com/rails/rails/pull/52951/commits/164e19c1d461da9759b9356c8758205533733974

end
```

This check is performed only if `terms_of_service` is not `nil`. The default

> error message for this helper is _"must be accepted"_. You can also pass in a

Reviewer 3: ```suggestion
error message for this validation is _"must be accepted"_. You can also pass in a
```

The default error message is _"can't be blank"_.

[`Object#blank?`]:
    https://api.rubyonrails.org/classes/Object.html#method-i-blank-3F

> ### `absence`

Reviewer 3: These validations seems to be sorted alphabetically, except for `absence`.
Ah, it seems `inclusion` and `exclusion` as seperate sections also broke this pattern.
No specific thing to change here I guess, just wanted to make a note.

Author: Since `absence` is a standalone section and does not need to be aside `presence` I've moved `absence` to the top of the Validations section.  It also has a link to `presence` which helps `absence` find its partner. See https://github.com/rails/rails/pull/52951/commits/7edd0b809e536008624b8e28ef355f92082c2f1d

| `:equal_to`                 | Specifies the value must be equal to the supplied value.                 | "must be equal to %{count}"                 |
| `:less_than`                | Specifies the value must be less than the supplied value.                | "must be less than %{count}"                |
| `:less_than_or_equal_to`    | Specifies the value must be less than or equal to the supplied value.    | "must be less than or equal to %{count}"    |
| `:other_than`               | Specifies the value must be other than the supplied value.               | "must be other than %{count}"               |
| `:in`                       | Specifies the value must be in the supplied range.                       | "must be in %{count}"                       |
| `:odd`                      | Specifies the value must be an odd number.                               | "must be odd"                               |
| `:even`                     | Specifies the value must be an even number.                              | "must be even"                              |

### `presence`

> This helper validates that the specified attributes are not empty. It uses the

Reviewer 2: I suppose that the following should be described first for newbies:

```diff
- This helper validates that the specified attributes are not empty.
+ This helper validates that the specified attributes or an association are not empty.
```

I suppose that the following (passing an association and attributes simultaneously) should also be clarified for newbies:

```ruby
class Supplier < ApplicationRecord
  has_one :account
  validates :account, :name, :login, :email, presence: true # possible? impossible? unrecommended?
end
```

Author: @hachi8833 thanks so much for your review!

> I suppose that the following should be described first for newbies:
> This helper validates that the specified attributes or an association are not empty.

@hachi8833 I don't think we need to be so explicit on this because we mention the associations in the second paragraph. 

> I suppose that the following (passing an association and attributes simultaneously) should also be clarified for newbies:
> 
> ```
> class Supplier < ApplicationRecord
>   has_one :account
>   validates :account, :name, :login, :email, presence: true # possible? impossible? unrecommended?
> end
> ```
> 

This is possible, but @p8 do you know anything about whether it's recommended?

My gut feeling is that to maintain clarity and improve readability, it’s better to separate
validations for associations and individual attributes. If there's no recommendations then I'm not sure that we need to outline it.

Reviewer 2: Thank you for looking into them!

> I don't think we need to be so explicit on this because we mention the associations in the second paragraph.

OK, I leave it to you 🙏

Reviewer 3: If the attributes list is long and a association is added in the middle it's probably easy to miss 🤷‍♂️ 
I'd think the recommend implementation would be passing `required: true` to the association:
```ruby
has_one :account, required: true
```
We could mention that as well?

There also is the validate option that we might need to mention?
<img width="975" alt="image" src="https://github.com/user-attachments/assets/0af7cd52-2063-402a-96b9-20efaa62af30">
https://edgeapi.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-belongs_to

Author: > We could mention that as well?

If nobody feels strongly about this I think I'm going to defer mentioning this for now. I think this may convolute the section a bit. 

> There also is the validate option that we might need to mention?

I did however add this section https://github.com/rails/rails/pull/52951/commits/6ccc4d173a902194464aa94e657bc7b71eb630b6

> This helper validates that the specified attributes are not empty. It uses the

Reviewer 3: ```suggestion
This validator validates that the specified attributes are not empty. It uses the
```

| `:less_than_or_equal_to`    | Specifies the value must be less than or equal to the supplied value.    | "must be less than or equal to %{count}"    |
| `:other_than`               | Specifies the value must be other than the supplied value.               | "must be other than %{count}"               |
| `:in`                       | Specifies the value must be in the supplied range.                       | "must be in %{count}"                       |
| `:odd`                      | Specifies the value must be an odd number.                               | "must be odd"                               |
| `:even`                     | Specifies the value must be an even number.                              | "must be even"                              |

### `presence`

This validator validates that the specified attributes are not empty. It uses the
[`Object#blank?`][] method to check if the value is either `nil` or a blank
string - that is, a string that is either empty or consists of whitespace.

```ruby
class Person < ApplicationRecord
  validates :name, :login, :email, presence: true
end
```

```irb

> person = Person.new(name: "Alice", login: "alice123", email: "alice@example.com")
> person.valid?

Reviewer 3: ```suggestion
irb> person = Person.new(name: "Alice", login: "alice123", email: "alice@example.com")
irb> person.valid?
```

=> true # presence validation passes

> invalid_person = Person.new(name: "", login: nil, email: "bob@example.com")
> invalid_person.valid?

Reviewer 3: ```suggestion
irb> invalid_person = Person.new(name: "", login: nil, email: "bob@example.com")
irb> invalid_person.valid?
```

end
```

NOTE: If you want to ensure that the association is both present and valid, you
also need to use `validates_associated`. More on that in the
[validates_associated section](#validates-associated).

If you validate the absence of an object associated via a
[`has_one`](association_basics.html#the-has-one-association) or
[`has_many`](association_basics.html#the-has-many-association) relationship, it
will check that the object is neither `present?` nor `marked_for_destruction?`.

> Since `false.present?` is false, if you want to validate the absence of a
> boolean field you should use `validates :field_name, exclusion: { in: [true,
> false] }`.

Reviewer 3: I think the example here is a bit more readable if we use the triple backticks:
````suggestion
Since `false.present?` is false, if you want to validate the absence of a
boolean field you should use:

```ruby
validates :field_name, exclusion: { in: [true, false] }
```
````

end
```

```irb

> account = Account.create(name: "Account A")

Reviewer 3: ```suggestion
irb> account = Account.create(name: "Account A")
```

> supplier = Supplier.new(account: account)
> supplier.valid?

Reviewer 3: ```suggestion
irb> supplier = Supplier.new(account: account)
irb> supplier.valid?
```

=> true # presence validation passes

> invalid_supplier = Supplier.new
> invalid_supplier.valid?

Reviewer 3: ```suggestion
irb> invalid_supplier = Supplier.new
irb> invalid_supplier.valid?
```

end
```

WARNING: Some databases are configured to perform case-insensitive searches
anyway.

> A `:conditions` option can be used to specify additional conditions as a `WHERE`
> SQL fragment to limit the uniqueness constraint lookup (e.g. `conditions: -> {
> where(status: 'active') }`).

Reviewer 3: I think the example here is a bit more readable if we use the triple backticks and a full example:

````suggestion
A `:conditions` option can be used to specify additional conditions as a `WHERE`
SQL fragment to limit the uniqueness constraint lookup:

```ruby
validates :name, uniqueness: { conditions: -> { where(status: 'active') } }
```
````

[`add_index`]:
    https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_index
[the MySQL manual]:
    https://dev.mysql.com/doc/refman/en/multiple-column-indexes.html
[the MariaDB manual]: https://mariadb.com/kb/en/compound-composite-indexes/
[the PostgreSQL manual]:
    https://www.postgresql.org/docs/current/static/ddl-constraints.html

### `validate`

The `validate` validator is used with `belongs_to` and `has_one` associations to
automatically validate the associated object whenever the parent object is
saved. This option ensures that any new or updated associated objects are
validated before the parent object can be successfully saved. By default, this
option is set to `false`, meaning associated objects will not be automatically
validated.

```ruby
class Post < ApplicationRecord
  belongs_to :author, validate: true
end
```

In this example, every time a `Post` is saved, the associated `Author` will also
be validated. If the `Author` is invalid, the `Post` will not be saved.

If you need to ensure that associated objects are validated every time the

> parent object is updated, consider using `validates_associated`.

Reviewer 3: Argh, this option opens another can of worms 😬 

The `validate` options isn't really a validator, so I don't think it belongs here.
Also [the logic ](https://github.com/rails/rails/blob/ac98964910369dd9821316a68f5477bd0d8e4ced/activerecord/lib/active_record/reflection.rb#L700-L711)of when a record is/isn't validated is pretty complex 😞 
And autosave has [more corner cases](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/autosave_association.rb)

I don't think this is properly documented anywhere (both in the API docs and the guides), so I think it's good to have it documented somewhere. Maybe we can add another section to this guide? Or maybe it belongs in the associations guide?
I've added a rough outline.
````markdown

Automatic Association Validations
---------------------------------

Active Record enables validations on some associations.

### `belongs_to` Validations

By default `belongs_to` associations validate the presence of the associated
record to guarantee reference consistency every time the record is saved.

```ruby
class Post < ApplicationRecord
  belongs_to :author
end
```

```irb
irb> post = Post.new
irb> post.valid?
=> false # Author is missing
```

Set the `:optional` option to true if this is not what you want. This means
that the foreign key in one table might not reliably point to a valid primary
key in the referenced table.

```ruby
class Post < ApplicationRecord
  belongs_to :author, optional: true
end
```

### `has_many` Validations

For `has_many` and `has_many :through` associations the associated records will
be validated every time the parent is saved.

```
class Post < ActiveRecord::Base
  belongs_to :author
  validates :title, presence: true
end

class Author < ActiveRecord::Base
  has_many :posts
end
```

```irb
irb> author = Author.new
irb> post = author.posts.new
irb> author.save
=> false
irb> post.title = "Some title"
irb> author.save
=> true

### `validate` Option

The `validate` option is used with `belongs_to` and `has_one` associations to
automatically validate the associated object whenever the parent object is
saved. This option ensures that any new associated objects are
validated before the parent object can be successfully saved. By default, this
option is not set, meaning associated objects will not be automatically
validated.

```ruby
class Post < ApplicationRecord
  belongs_to :author, validate: true
end
```

In this example, every time a `Post` is saved, the associated `Author` will also
be validated. If the `Author` is invalid, the `Post` will not be saved.

If you need to ensure that associated objects are validated every time the
parent object is updated, consider using `validates_associated`.

### `autosave` Option

When the `autosave` option is set to true, the associated records will be
validated everytime the parent record is saved, instead of just when creating.

```ruby
class Supplier < ApplicationRecord
  has_one :account, autosave: true
end

class Account < ApplicationRecord
  validates :terms, presence: true
end
```

```irb
irb> supplier = Supplier.first
irb> supplier.account.terms = nil
irb> supplier.save
=> false # Account is invalid
```
````

Reviewer 4: Association guide makes sense

Author: @p8 as we discussed offline - looking at the section I want to try a table approach in the association guides which I think would be easier to digest but I suspect that it will need a few iterations so maybe a separate PR is better.  I've removed the original section here and created a task for it on our board.

If you want to add an error on a specific attribute, pass it as the first
argument, such as `record.errors.add(:first_name, "please choose another

> name")`. We will cover [validation errors](#working-with-validation-errors) in greater detail later.

Reviewer 3: I think the inline example here can be removed, as we already show a complete example next.

```

> You can also use `on:` to define custom contexts. Custom contexts need to be

Author: Oops!

```

This means that any field with an error ends up with a 2 pixel red border.

### Customizing Error Field Wrapper

Rails uses the `field_error_proc` configuration option to wrap fields with
errors in HTML. By default, this option wraps the erroneous form fields in a
`<div>` with a `field_with_errors` class, as seen in the example above:

```ruby
config.action_view.field_error_proc = Proc.new { |html_tag, instance| content_tag :div, html_tag, class: "field_with_errors" }
```

You can customize this behavior by modifying the field_error_proc setting in
your application configuration, allowing you to change how errors are presented

> in your forms. For more details, refer to the [API documentation on
> field_error_proc](https://api.rubyonrails.org/configuring.html#config-action-view-field-error-proc).

Reviewer 3: This link seems broken. It should probably be the following (although it doesn't add more information):
```suggestion
in your forms. For more details, refer to the [Configuration Guide on
field_error_proc](configuring.html#config-action-view-field-error-proc).
```
